### ISSUE TITLE
make test very slow on Windows

### ISSUE THREAD

**Issue by bbbradsmith** [cc65/cc65#2095]
During investigation of #2092 it was found that "make test" runs about 5x slower on the Github action builds for Windows than Linux. I'm not sure of the cause.

It could be that the MSVC builds are really that much slower, in which case I think there must be some incorrect compiler settings, as there really should not be a 5x speedup between this and MinGW GCC.

The other stronger suspicion I have is that **"make" is somehow much slower on the Windows platform**. If that's the case I'm not sure if there's much we can do about it, unless there's something problematic about the way our particular makefiles are written that is an issue for Windows only.

Either way, it would be nice to understand it, and even nicer to fix it.

Edit: for quick reference I've made the correct answer **bold** above. (The binaries are fine, make is the problem.)

**Comment by mrdudz** [comment]
> The other stronger suspicion I have is that "make" is somehow much slower on the Windows platform. 

That is the case. Windows is awfully slow at spawning processes (which happens a lot during make). I don't think there is a way to fix this. VICE builds really really slow on Windows for that matter (for the same reasons).

That said, the build environment doesn't seem to ever truly use a native windows shell - at least not for the subshells used by "make". see #1920 and #1937 for that matter

**Comment by bbbradsmith** [comment]
So I made a test on my fork, building and testing both MSVC and GCC in the same job (just to be sure it's the same machine at least), and then another job in the reverse order. The time differences are not significant between the two, so I think that at least puts to rest the idea that the MSVC is horribly misconfigured. Details follow, if curious:

# Details

Starting from #2092 which fixes sim65 so that the MSVC test can pass.

I created [a workflow](https://github.com/bbbradsmith/cc65/blob/0749a59968af620b29f439b8d43fd30b46399702/.github/workflows/windows-performance-test.yml) that looked like this:
```
name: Windows Performance Test
on: [pull_request]

jobs:

  build_msvc_gcc:
    name: Build MSVC then GCC
    runs-on: windows-latest

    steps:
      - name: Git Configure
        shell: bash
        run: git config --global core.autocrlf input
      - name: Checkout Source
        uses: actions/checkout@v3

      - name: MSVC Add msbuild to PATH
        uses: microsoft/setup-msbuild@v1.1
      - name: MSVC Build app (release)
        run: msbuild src\cc65.sln -t:rebuild -property:Configuration=Release
      - name: MSVC make lib
        run: make -j2 lib QUIET=1
      - name: MSVC make test
        run: make test QUIET=1

      - name: Blank Slate
        shell: bash
        run: ls -A1 | xargs rm -rf
      - name: Git Configure
        shell: bash
        run: git config --global core.autocrlf input
      - name: Checkout Source
        uses: actions/checkout@v3

      - name: MinGW Build app
        run: make -j2 bin USER_CFLAGS=-Werror
      - name: MinGW make lib
        run: make -j2 lib QUIET=1
      - name: MinGW make test
        run: make test QUIET=1

  build_gcc_msvc:
    name: Build GCC then MSVC
    runs-on: windows-latest

    steps:
      - name: Git Configure
        shell: bash
        run: git config --global core.autocrlf input

      - name: Checkout Source
        uses: actions/checkout@v3

      - name: MinGW Build app
        run: make -j2 bin USER_CFLAGS=-Werror
      - name: MinGW make lib
        run: make -j2 lib QUIET=1
      - name: MinGW make test
        run: make test QUIET=1

      - name: Blank Slate
        shell: bash
        run: ls -A1 | xargs rm -rf
      - name: Git Configure
        shell: bash
        run: git config --global core.autocrlf input
      - name: Checkout Source
        uses: actions/checkout@v3

      - name: MSVC Add msbuild to PATH
        uses: microsoft/setup-msbuild@v1.1
      - name: MSVC Build app (release)
        run: msbuild src\cc65.sln -t:rebuild -property:Configuration=Release
      - name: MSVC make lib
        run: make -j2 lib QUIET=1
      - name: MSVC make test
        run: make test QUIET=1
```

The [timing results](https://github.com/bbbradsmith/cc65/actions/runs/4896224581):
```
Build MSVC then GCC
 00:01 - Set up job
 00:00 - Git Configure
 00:07 - Checkout Source
 00:01 - MSVC Add msbuild to PATH
 01:44 - MSVC Build app (release)
 01:54 - MSVC make lib
 27:14 - MSVC make test
 00:05 - Blank Slate
 00:00 - Git Configure
 00:03 - Checkout Source
 00:44 - MinGW Build app
 01:28 - MinGW make lib
 24:19 - MinGW make test
 00:01 - Post Checkout Source
 00:01 - Post Checkout Source
 00:00 - Complete job
 --------------------
 57:47 - Total


Build GCC then MSVC
 00:01 - Set up job
 00:02 - Git Configure
 00:07 - Checkout Source
 01:09 - MinGW Build app
 01:56 - MinGW make lib
 30:07 - MinGW make test
 00:06 - Blank Slate
 00:00 - Git Configure
 00:04 - Checkout Source
 00:01 - MSVC Add msbuild to PATH
 01:24 - MSVC Build app (release)
 02:17 - MSVC make lib
 29:32 - MSVC make test
 00:01 - Post Checkout Source
 00:01 - Post Checkout Source
 00:00 - Complete job
 --------------------
 60:06 - Total
 ```

**Comment by bbbradsmith** [comment]
> That is the case. Windows is awfully slow at spawning processes (which happens a lot during make). I don't think there is a way to fix this. VICE builds really really slow on Windows for that matter (for the same reasons).

> That said, the build environment doesn't seem to ever truly use a native windows shell - at least not for the subshells used by "make". see #1920 and #1937 for that matter

I've had make work from cmd on some projects, but it's always been a pain trying to figure out how, with platform-specific substitutions needed. My normal approach is just to use MSYS2 whenever I want to run make. I'll try another test that tries all 3.

One thing I've noticed locally is that if I have one of the source directories open in an explorer window (e.g. test/val) then the explorer itself starts using a lot of CPU while make is building files out of it. That seemed very odd to me, and annoying... I can make faster if I close the unchanging source folder while it's running??

**Comment by mrdudz** [comment]
It boils down to: Windows being Windows =)

**Comment by bbbradsmith** [comment]
So, I'm [running a test](https://github.com/bbbradsmith/cc65/actions/runs/4897404590) that [compares all 3 windows shells](https://github.com/bbbradsmith/cc65/actions/runs/4897404590/workflow). All 3 took and equivalent amount of time.

I'm surprised to say that all 3 make steps that I tested reported no problems with ```shell: cmd```?
```
        run: make -j2 bin USER_CFLAGS=-Werror
        run: make -j2 lib QUIET=1
        run: make test QUIET=1
```
What was blocking #1920 or #1937 in the past?

**Comment by mrdudz** [comment]
try doing something from inside a make rule that actually requires cmd.exe - make itself IS run from cmd, but the subshell make uses isnt cmd

**Comment by bbbradsmith** [comment]
Well so far all runner makes with shell: cmd look fine and I don't see any evidence of it escaping the shell. I've now started a local cmd.exe make, and I'll see what happens there. So far libs are building fine...

**Comment by mrdudz** [comment]
local it works as expected (now). back then it did not, and i tried making the test verify it - however, apparently the sub shells are some WSL thing, not cmd, which made it not work as expected at all. The makefiles do a lot of magic to support cmd - which is what i wanted to test. if those things are not actually used, there is very little point in running them twice :)

**Comment by bbbradsmith** [comment]
I don't have WSL installed. I do have MSYS2 but I don't think it's possible for my CMD make operation to make use of MSYS2.

```make -j2 libs``` worked.
```make -j2 samples``` worked
```make test``` is having trouble with the clean, starts off with a bunch of errors like this...
```
The system cannot find the path specified.
make[3]: [clean] Error 3 (ignored)
The system cannot find the path specified.
make[3]: [clean] Error 3 (ignored)
...
```
The runners have a lot of things in their environment I don't, but I'll figure this out shortly...

**Comment by mrdudz** [comment]
Again: it works locally. Its about what the github runners do. You cant reproduce this locally.

**Comment by bbbradsmith** [comment]
Well, my local build problem is just not having GCC to build ISEQUAL which is done in several of the test makefiles.

I'm unclear what you're saying the github runners do that I wouldn't be able to reproduce locally. You mentioned that they might open a shell other than cmd, but how can we spot this happening? I can't find any evidence of such a thing in the build tests I've been doing so far.

At this point the runner appears to be happy to do almost the whole set of tests in cmd so far, libs, util, tests are all passing, ony samples remains, because of a [weird issue with the geos sample](https://github.com/cc65/cc65/pull/2092#issuecomment-1536868347) which I'm in the midst of diagnosing, but that doesn't sound like what you were referring to?

**Comment by bbbradsmith** [comment]
Anyhow, with #2098 and #2099 I find that the Windows make test succeeds in cmd for me locally (with make and gcc on the path), and on the runners with #2101. I've tried all 3 available shell options for the make but they have equivalent speed.

I noticed we had -j2 on all the other makes in the existing linux build and test, however `make test` does not, and appears to fail if -j is used.

**Comment by mrdudz** [comment]
Let me try to explain again:

The point of #1920 and #1937 is adding a workflow that tests what happens when you use the makefiles locally on a windows cmd.exe shell. The point is NOT to run the makefiles or tests "on windows" (that works fine, and is out of scope).

Now, when i tried doing that in #1937 it turned out that the github runner doesnt actually use cmd.exe as its main environment even when configured to use cmd.exe - instead it uses some other environment (my guess was WSL, i couldnt really find out what), and then uses something like ```cmd /C <command>``` to run individual commands from the GHA. And then ```make``` would - for whatever reason - not really run in a proper cmd.exe environment, and when it runs commands itself, those would *not* run on cmd.exe, but something else. So as a result you aren't reproducing what happens when you run the "make" on your local cmd.exe shell, but something different, which does not help in actually confirming that everything works fine on a cmd.exe shell. (it does still "work" - but thats besides the point)

However, i am closing this - since "runs slow" is just what it is.

**Comment by bbbradsmith** [comment]
Could you elaborate with something specific about what's different?

I am really trying hard to understand what difference you're referring to, but I am unable to follow what problem you're describing.

This was the closest I understood:

> try doing something from inside a make rule that actually requires cmd.exe - make itself IS run from cmd, but the subshell make uses isnt cmd. Or well, wasnt when i did those tests. (The point was: test if those things actually work on cmd, which they didnt at some point before)

However, I don't know what "something" to do to verify this. What do I need to do in a build to check whether this is the case?

I looked closely at the build logs, but I can't see any evidence that anything besides cmd is being used where requested.

With #2092 I am able to fully build all tests and samples on my local cmd (requiring make and gcc), and also in the created build action. Can you tell me something I can test to demonstrate the difference you're talking about? Is there a way to see it in the logs that I'm missing?

**Comment by mrdudz** [comment]
> I am really trying hard to understand what difference you're referring to, but I am unable to follow what problem you're describing.

I honestly don't know what else to explain. The point is that if the subshell used by make isnt actually cmd, then we are not testing what we want to test. Even if it "works" it doesn't actually tell you "it works". It only tells "it works in the GHA on the github runner".

> However, I don't know what "something" to do to verify this. What do I need to do in a build to check whether this is the case?

I don't remember what exactly was the problem back then - but "try running a cmd specific batchfile" is the first thing that comes to mind. Anything that would work on cmd but not in bash could show it.

**Comment by bbbradsmith** [comment]
> "try running a cmd specific batchfile" is the first thing that comes to mind. Anything that would work on cmd but not in bash could show it.

That's what I'm asking for a specific example of. Could you spell out how to do that? I would be happy to test this but I don't know how to go about creating it.

**Comment by mrdudz** [comment]
took me a bit of thinking, this should be a simple test: in a Makefile rule, use the command
```
echo off
```
In bash, this prints "off". In cmd it prints nothing, but instead disables echoing of commands.

**Comment by bbbradsmith** [comment]
Ah I see. I tried that and it does indeed echo the off on the runner test.

I found that with `make -d` we can see this:
```find_and_set_shell() path search set default_shell = C:/Program Files/Git/bin/sh.exe"```

This seems to happen regardless of the SHELL environment variable.

**Comment by bbbradsmith** [comment]
Okay, progress. Using ```set SHELL=cmd``` (or other variations, or the explicit path) did nothing but... feeding it directly to make does the trick:
```make test QUIET=1 SHELL=cmd```

This passes the echo test. It also [ran about 10 minutes faster](https://github.com/bbbradsmith/cc65/actions/runs/4910958688)?? [Workflow YML](https://github.com/bbbradsmith/cc65/actions/runs/4910958688/workflow). With `-d` it did confirm using cmd as requested, and shows how it builds a ton of temporary batch files as part of the execution (which my local cmd does the same).

The only issue is that `make -j2 lib QUIET=1 SHELL=cmd` failed, but without `-j2` it was fine. That's a little bit curious.

**Comment by bbbradsmith** [comment]
A [speed test of cmd vs pwsh vs sh](https://github.com/bbbradsmith/cc65/actions/runs/4911160202) shows that cmd is consistently about 10 minutes faster for `make test` than the other 2. (Though I need to verify that pwsh isn't actually sh.)
